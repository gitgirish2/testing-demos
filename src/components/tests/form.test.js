import React from "react";
import {shallow} from "enzyme"
import LoginForm from "../form"
import InputComponent from "../lib/input"
import { Button } from '@material-ui/core';


describe('LoginForm', () => {
    let wrapper
    beforeEach(() => {
        wrapper = shallow(<LoginForm />)
    })

    it("should render", () => {
        expect(wrapper).toMatchSnapshot()
    })

    it("should render with correct initial state", () => {
        // arrange
        const expectedInitialState = {
           username: {value: '', error: true, message: ''},
            password: {value: '', error: true, message: ''},
            error: false,
            errorMessage: '',
            isLoginEnabled: false
        }

        // act
        const state = wrapper.state()

        // assert
        expect(state).toEqual(expectedInitialState)
    })

    it("should have disabled login button initially", () => {
        // arrange
        const isDisabled = true

        // act
        const loginBtn = wrapper.find(Button)

        // assert
        expect(loginBtn.props().disabled).toEqual(isDisabled)
    })

    it("should have updated username in state when username is changed", () => {
        // arrange
        const username = {value: 'user', error: false, message: ''}

        // act
        const UsernameInput = wrapper.find(InputComponent).at(0)
        UsernameInput.props().onChange('user')

        // assert
        expect(wrapper.state().username).toEqual(username)
    })

    it("should update password in state when password is changed", () => {
        // arrange
        const password = {value: 'password', error: false, message: ''}

        // act
        const PasswordInput = wrapper.find(InputComponent).at(1)
        PasswordInput.props().onChange('password')

        // assert
        expect(wrapper.state().password).toEqual(password)
    })

    it("should have enabled login button after filling username and password", () => {
        // arrange
        const isDisabled = false

        // act
        const UsernameInput = wrapper.find(InputComponent).at(0)
        const PasswordInput = wrapper.find(InputComponent).at(1)

        UsernameInput.props().onChange('user')
        PasswordInput.props().onChange('password')

        const loginBtn = wrapper.find(Button)

        // assert
        expect(loginBtn.props().disabled).toEqual(isDisabled)
    })

    it("should show error on login if password is not valid", () => {

        // act
        const UsernameInput = wrapper.find(InputComponent).at(0)
        const PasswordInput = wrapper.find(InputComponent).at(1)

        UsernameInput.props().onChange('user')
        PasswordInput.props().onChange('pass')

        const loginBtn = wrapper.find(Button)
        loginBtn.props().onClick()

        // assert
        expect(wrapper.state().error).toEqual(true)
        expect(wrapper.state().errorMessage).toEqual("The password should be at least 6 characters long")
    })
})