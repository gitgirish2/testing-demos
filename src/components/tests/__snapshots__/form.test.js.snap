// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm should render 1`] = `
<Fragment>
  <WithStyles(ForwardRef(FormHelperText))
    error={false}
    id="form-error-text"
  />
  <form
    className="login-form"
  >
    <InputComponent
      id="username"
      inputErrorDetails={
        Object {
          "error": true,
          "message": "Username is empty",
          "value": "",
        }
      }
      label="Username"
      name="username"
      onChange={[Function]}
      placeholder="Username"
      value=""
    />
    <InputComponent
      id="password"
      inputErrorDetails={
        Object {
          "error": true,
          "message": "Password is empty",
          "value": "",
        }
      }
      label="Password"
      name="password"
      onChange={[Function]}
      placeholder="Password"
      value=""
    />
    <WithStyles(ForwardRef(Button))
      color="primary"
      disabled={true}
      onClick={[Function]}
      variant="contained"
    >
      Login
    </WithStyles(ForwardRef(Button))>
  </form>
</Fragment>
`;
